OKR.md

```diff
@@ -0,0 +1,50 @@
+# Quarterly Objectives and Key Results
+
+We try to frame our ongoing work using a process based on quarterly Objectives and Key Results (OKRs). Objectives reflect outcomes that are challenging, but realistic. Results are tangible and measurable.
+
+## 2018 Q2
+
+> Work in Progress
+
+### The daemon is stable and so core is reliable
+
+- `PX` - OWNER - The uncaught errors and memory usage are under control and a daemon can run for a week non stop
+- `PX` - OWNER - One or more js-ipfs daemons are part of the IPFS infrastructure
+- `PX` - OWNER - js.ipfs.io exposes an IPFS Gateway provided by a js-ipfs daemon
+
+### The core can handle vast amounts of data
+
+- `PX` - OWNER - The MFS is implemented in js-ipfs
+- `PX` - OWNER - npm on IPFS is used as a test vector of js-ipfs capabilities
+- `PX` - OWNER - js-ipfs in the browser can load a dataset of at least 100Gb (create archives.js.ipfs.io)
+- `PX` - OWNER - js-ipfs in Node.js can load a dataset of at least 1TB (use npm to prove it)
+- `PX` - OWNER - A S3 IPFS Repo backend exists and it is easily pluggable on `new IPFS`
+- `PX` - OWNER - The PIN API is finished and merged.
+
+### The APIs are more ergonomic and developer productivity is off the charts
+
+- `PX` - OWNER - Complete the first milestone of the Awesome Endeavour - Flow
+- `PX` - OWNER - Roadmap and Kickstart the second milestone for the Awesome Endeavour - Flow
+- `PX` - OWNER - Make js-ipfs uglify friendly again
+- `PX` - OWNER - Every env test run is parallelized in CI
+- `PX` - OWNER - CLI, HTTP-API and Core Tests of js-ipfs are parallelized in CI
+- `PX` - OWNER - A repo migration utility exists and migration is a good experience for js-ipfs users
+
+### libp2p levels up, learns new superpowers and a dedicated team is formed
+
+- `PX` - OWNER - libp2p has primitives to keep connections under control
+- `PX` - OWNER - Readable Streams no more! Reduce the overhead of Readable -> Pull in libp2p-mplex
+- `PX` - OWNER - The libp2p.next() refactor is complete
+- `PX` - OWNER - Delegated Peer Routing and Content Routing modules are created
+- `PX` - OWNER - Browsers no longer rely on static rendezvous points for Discovery
+- `PX` - OWNER - js-libp2p gets its own OKR spreadsheet for next Q3
+
```

Comment:
Can I propose the libp2p dissector for wireshark and libp2p-nodetrust as an OKR?
https://github.com/mkg20001/libp2p-dissector https://github.com/mkg20001/libp2p-nodetrust

Reply:
that be super helpful, not sure if it fits the js OKR tho, maybe it has a better place? In any case, would be rad to have it 👍

Reply:
@mkg20001 what would be the specific key result and would you be the owner of it?

Reply:
that's excellent @mkg20001. before writing them on the list, I just want to make sure that we are on the same page of the time commitment. Being the owner for both of the KR means that you have the time to allocate to both.

Reply:
Yes. Although I would probably need some help with the C stuff as I'm still learning the language.

Reply:
@diasdavid Are they going to get added to the list?

Reply:
@mkg20001 if you need help with the Wireshark dissector for libp2p, I'm happy to collaborate! I just finished working on a dissector for the Ethereum discovery protocol – which hopefully we'll publish soon ;-)

Reply:
@mkg20001 Is there a branch with the WIP you can point me to? Not sure if you've come across the Wireshark API facilities for reassembling PDUs over TCP, I guess yes [(section 2.7 here)](https://code.wireshark.org/review/gitweb?p=wireshark.git;a=blob_plain;f=doc/README.dissector).

I read the yamux spec and I guess another complexity you've identified is how to follow and register individual streams separately within a session. Since all Yamux packets contain the stream ID, we can set it on the proto tree and attach a filter to it.

Then I think we can use the `register_follow_stream` function to register a new follower on a statistics tap, that relies on the stream ID. This function is somewhat hidden, as I wasn't able to find documentation about this facility, but the tcp dissector [uses it here] (https://github.com/wireshark/wireshark/blob/4c14ac075403e3f5ea5709501918c08e45272439/epan/dissectors/packet-tcp.c#L7619)

Perhaps sooner or later, we'll need to store the streams identified within a session inside the WS conversation data. We can either use a simple linked list of stream IDs, or a map if we want to store info about each stream (e.g. statistics).

Happy to continue discussing, here or elsewhere ;-)